File: Medium/Shortest path in Directed Acyclic Graph/shortest-path-in-directed-acyclic-graph.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++
class Solution {
 private:
    void dfs(map<int,list<pair<int,int> >> &adjlist, stack<int>&topo,vector<bool> &visited,int v)
    {
        visited[v]=true;
        
        for(auto x : adjlist[v])
        {
            int vt = x.first;
            int wt = x.second;
            
            if(visited[vt]!=true)
            {
                dfs(adjlist,topo,visited,vt);
            }
        }
        topo.push(v);
        return;
    }
  public:
     vector<int> shortestPath(int N,int M, vector<vector<int>>& edges){
        // code here
        // vector<int> dis(N,-1);
        
        //create adjlist
        map<int,list<pair<int,int> >> adjlist;
        for(auto x: edges)
        {
            int i = x[0];
            int j = x[1];
            int w = x[2];
            
            adjlist[i].push_back({j,w});
        }
        stack<int> topo;
        vector<bool> visited(N,false);
        dfs(adjlist,topo,visited,0);
        vector<int> dis(N,-1);
        dis[0]=0;
       while(!topo.empty())
       {
           int front = topo.top();
           topo.pop();
           for(auto x : adjlist[front])
           {
               int v = x.first;
               int w = x.second;
               if(dis[v]==-1)
               {
                   dis[v]=dis[front]+w;
               }
               if(dis[v]> dis[front]+w)
               {
                   dis[v]=dis[front]+w;
               }
           }
           
       }
       return dis;
    }
};


//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, m;
        cin >> n >> m;
        vector<vector<int>> edges;
        for(int i=0; i<m; ++i){
            vector<int> temp;
            for(int j=0; j<3; ++j){
                int x; cin>>x;
                temp.push_back(x);
            }
            edges.push_back(temp);
        }
        Solution obj;
        vector<int> res = obj.shortestPath(n, m, edges);
        for (auto x : res) {
            cout << x << " ";
        }
        cout << "\n";
    }
}

// } Driver Code Ends

File: Difficulty: Medium/Shortest path in Directed Acyclic Graph/shortest-path-in-directed-acyclic-graph.py
#User function Template for python3

from typing import List
import math


class Solution:

    def shortestPath(self, V: int, E: int, edges: List[List[int]]) -> List[int]:
        
        adjList ={i:[] for i in range(V)}
        visited = [0 for i in range(V)]
        
        #creating the adjlist
        for i in range(len(edges)):
            u,v,w = edges[i]
            adjList[u].append([v,w])
            
        
        #DFS toposort
        topo=[]
        def dfs_topo(start):
            visited[start]=1
            
            for v,w in adjList[start]:
                
                if visited[v]==0:
                    
                    dfs_topo(v)
            
            topo.append(start)
       
                  
        dfs_topo(0)

        
        distance = [-1 for i in range(V)]
        distance[0] = 0
        
        while topo:
            u = topo[-1]
            topo.pop()
            
            for nei in adjList[u]:
                v,w  = nei
                
                if(distance[u]+w < distance[v] or distance[v]==-1):
                    distance[v] = distance[u]+w
            
            
        
        return distance
        
        



#{ 
 # Driver Code Starts
#Initial Template for Python 3

from typing import List


class IntMatrix:

    def __init__(self) -> None:
        pass

    def Input(self, n, m):
        matrix = []
        #matrix input
        for _ in range(n):
            matrix.append([int(i) for i in input().strip().split()])
        return matrix

    def Print(self, arr):
        for i in arr:
            for j in i:
                print(j, end=" ")
            print()


class IntArray:

    def __init__(self) -> None:
        pass

    def Input(self, n):
        arr = [int(i) for i in input().strip().split()]  #array input
        return arr

    def Print(self, arr):
        for i in arr:
            print(i, end=" ")
        print()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):

        n, m = map(int, input().split())

        edges = IntMatrix().Input(m, 3)

        obj = Solution()
        res = obj.shortestPath(n, m, edges)

        IntArray().Print(res)
        print("~")

# } Driver Code Ends